### PR TITLE
Test all language generators in one command

### PR DESCRIPTION
**Pull request by jhowlett-scottlogic**
PR related to issue  https://github.com/ScottLogic/openapi-forge/issues/10

PR related to PR https://github.com/ScottLogic/openapi-forge-typescript/pull/8

Using openapi-forge test-generators you can test all language generators and then have a table of results presented. 

This command searches for a local language generators or uses the version hosted on GitHub.

### CODE REVIEW

**Comment by jhowlett-scottlogic** [comment]
This approach uses a search and replace in some files (2 for TypeScript & 1 for CSharp) that refererences openapi-forge in the node_modules folder.  Instead these file paths are repalced to the local version of the generator and then reverted back. 

An alternative approach would be to have a separate configuration file in each language generator that would house these file paths (and another other variable data) and then the language generators will have a different configuration file when using the new test-generators  command compared to using the testing that is in each of the language generators.

A small refactoring to help prevent this testing from breaking if the file path lines change causing the regex patterns to start failing. 

Let me know what you think and i can bring this change in.

File: README.md [hunk 1]
```diff
@@ -119,4 +119,26 @@ The following is a very high-level overview of the generation process:
...
 
-TODO: Elaborate, how to test generators etc ...
+TODO: Elaborate etc ...
+
+## Testing
+
+You can test openapi-forge on all of the language generators from one command:
+
+~~~
+% openapi-forge help test-generators
```
Comment (ColinEberhardt) [inline README.md:129]: Adding as a new command for the CLI is a great idea 👍

File: README.md [hunk 2]
```diff
@@ -119,4 +119,26 @@ The following is a very high-level overview of the generation process:
...
+Usage: openapi-generator test-generators [options]
+
+Test language specific generators.
+
+Options:
+  -g, --generators <gens>            Narrow down the generators to test. Each letter is a generator, combine letters to test multiple generators, options are:
+                                     c (CSharp), t (TypeScript) (default: "ct")
+  -c, --csharp <csharpPath>          Sets the location of the CSharp generator. Default is a directory called 'openapi-forge-csharp' in the same location as
```
Comment (ColinEberhardt) [inline README.md:137]: _aside_ should the forge have hard-coded references to the generators?

File: src/generate.js [hunk 3]
```diff
@@ -172,14 +162,14 @@ async function generate(schemaPathOrUrl, generatorPathOrUrl, options) {
       }
     }
 
-    if (schema && schema.components && schema.components.schemas) {
+    if (schema?.components?.schemas) {
```
Comment (ColinEberhardt) [inline src/generate.js:165]: much cleaner

File: src/testGenerators.js [hunk 4]
```diff
@@ -0,0 +1,205 @@
...
+    this.skipped = isNaN(skipped) ? 0 : skipped;
+    this.undefined = isNaN(undefined) ? 0 : undefined;
+    this.failed = isNaN(failed) ? 0 : failed;
+    this.time = time;
+}
+
+
+
+async function testGenerators(options) {
+    let temporaryFolder;
```
Comment (ColinEberhardt) [inline src/testGenerators.js:21]: having lots of mutable variables is a bit of a 'code smell'!

File: src/testGenerators.js [hunk 5]
```diff
@@ -0,0 +1,205 @@
...
+    const typescript = options.generators.includes("t");
+    const csharp = options.generators.includes("c");
+    if(!typescript && !csharp) {
+        throw new Error(
+            `No language to test. Please provide a language that you would like to test.`
+        );
+    }
+
+    shell.cd(__dirname, shellOptions);
+    if(typescript) {
```
Comment (ColinEberhardt) [inline src/testGenerators.js:40]: From a 30,000 foot view I can see that code of a similar shape is duplicated later in this function. It is also quite long, and as more language generators are added will need refactoring. A sign that it is worth investing some more time in re-structuring.

File: src/testGenerators.js [hunk 6]
```diff
@@ -0,0 +1,205 @@
...
+        try {
+            log.standard(`\n${log.bold}${log.underline}TypeScript${log.resetStyling}`);
+            featurePath = "../openapi-forge/features/*.feature";
+            basePath = "../../../openapi-forge/src/generate";
+            generatorPath = options.typescript;
+
+            log.standard(`Loading TypeScript generator from '${generatorPath}'`);
+
+            // Check for local Typescript generator.
+            if (!fs.existsSync(generatorPath)) {
```
Comment (ColinEberhardt) [inline src/testGenerators.js:51]: idea 💡 - this resolution process is something that is used twice in this function, but also, the generator itself has to perform this resolution. Could this be abstracted into a 'resolver', you give it a generator reference, it takes care of going away and finding it.

File: src/testGenerators.js [hunk 7]
```diff
@@ -0,0 +1,205 @@
...
+                    path.join(os.tmpdir(), "generator")
+                );
+                log.verbose(`Cloning generator from https://github.com/ScottLogic/openapi-forge-typescript.git to ${generatorPath}`);
+                shell.exec(`git clone https://github.com/ScottLogic/openapi-forge-typescript.git ${generatorPath}`, shellOptions);
+
+                // Get original paths
+                featurePath = path.relative(generatorPath, path.join(__dirname, "../features/*.feature")).replaceAll("\\", "/");
+                basePath = path.relative(path.join(generatorPath, "features/support"), path.join(__dirname, "../src/generate")).replaceAll("\\", "/");
+            }
+            log.verbose("Replacing path to .feature files");
```
Comment (ColinEberhardt) [inline src/testGenerators.js:69]: I'm really not sure why we need to change file paths? this feels quite fragile.

File: src/testGenerators.js [hunk 8]
```diff
@@ -0,0 +1,205 @@
...
+            log.verbose("Installing generator dependencies");
+            shell.exec(`npm install`, shellOptions);
+
+            log.standard("Starting tests");
+            const test = shell.exec(`npm run test`, shellOptions);
+            const stdout = test.stdout.split("\n");
+            const stdoutLength = stdout.length;
+            
+            // Format the output of the testing.
+            let result = stdout[stdoutLength-2].match(/^(\d+)m(\d+)\.\d+s/);
```
Comment (ColinEberhardt) [inline src/testGenerators.js:88]: regex is a super powerful tool, but hard to maintain. Please make sure that you add a brief comment that explains the regex.

File: src/testGenerators.js [hunk 9]
```diff
@@ -0,0 +1,205 @@
...
+
+            if(result) {
```
Comment (ColinEberhardt) [inline src/testGenerators.js:90]: It looks like this code is parsing the TypeScript / cuccumber output in order to locate the test results. All looks good, but perhaps refactor this into a seperate file? `TestResultsParser` perhaps?

File: src/testGenerators.js [hunk 10]
```diff
@@ -0,0 +1,205 @@
...
+                let time = "";
+                if((result[1] !== "0") && (result[1] !== "")) time = `${result[1]}m`;
+                time = `${time}${result[2]}s` 
+
+                result = stdout[stdoutLength-4].match(/^(\d+)\sscenarios?\s\(((\d+)\sfailed)?(,\s)?((\d+)\sundefined)?(,\s)?((\d+)\spassed)?\)/);
+                
+                resultArray.TypeScript = new Result(parseInt(result[1]), parseInt(result[9]), 0, parseInt(result[6]), parseInt(result[3]), time);
+                }
+
+                if(!temporaryFolder) {
```
Comment (ColinEberhardt) [inline src/testGenerators.js:100]: See comment about generator resolving - could that also have responsibility for maintaining paths?

File: src/testGenerators.js [hunk 11]
```diff
@@ -0,0 +1,205 @@
...
+            log.standard(`              TypeScript testing ${log.redBackground}${log.blackForeground} FAILED ${log.resetStyling}`);
+            log.standard(`${log.divider}`);
+            if(log.isStandard()) {
+              log.standard(`${exception.message}`);
+            } else {
+              log.verbose(`${exception.stack}`);
+            }
+            log.standard(`${log.divider}`);
+        } finally {
+            if (temporaryFolder) {
```
Comment (ColinEberhardt) [inline src/testGenerators.js:119]: clean-up could also be a job for the resolver?

**Review by ColinEberhardt — CHANGES_REQUESTED** [review]
functionally works very well, but could do with a bit more refactoring.

File: src/testGenerators.js (hunk 4 above)
Comment (jhowlett-scottlogic) [reply src/testGenerators.js]: increased immutability

File: src/testGenerators.js (hunk 5 above)
Comment (jhowlett-scottlogic) [reply src/testGenerators.js:70]: I have extracted duplicated logic to separate functions, generatorResolver.js or testResultParser.js

File: src/testGenerators.js (hunk 7 above)
Comment (jhowlett-scottlogic) [reply src/testGenerators.js]: Now uses cl arguments to configure the generators

File: src/testGenerators.js (hunk 8 above)
Comment (jhowlett-scottlogic) [reply src/testGenerators.js]: I have added a small commnet above Regex uses to help explain them

File: src/testGenerators.js (hunk 9 above)
Comment (jhowlett-scottlogic) [reply src/testGenerators.js]: done

File: src/testGenerators.js (hunk 10 above)
Comment (jhowlett-scottlogic) [reply src/testGenerators.js]: generatorResolver.js handles getting and disposing of generators

File: src/testGenerators.js (hunk 6 above)
Comment (jhowlett-scottlogic) [reply src/testGenerators.js]: done

**Comment by jhowlett-scottlogic** [comment]
Changes since previous review:

- Handle new cl arguments in generator's `npm test`
- Extract parsing test results to testResultParser.js
- Extract generator download ans cleanup to generatorResolver.js
- Extract repeated logic in testGenerators()
- House shellOptions in log.js for one unified shellOptions

**Review by ColinEberhardt — APPROVED** [review]

**Pull request merged**